lib/ProductOpener/Import.pm

```diff
@@ -412,6 +412,167 @@ my %no = (
 );
 
 
+sub apply_organization_settings_to_product($$$$) {
+
+	my %stats = shift;
```

Comment:
hashes need to be passed by references, so it should be my $stats_ref, in the function should be called with \%stats as a parameter.

Reply:
Why would we change it? It works fine as is. 

And when I try to change it I get this error:
```
[Tue Aug  2 14:43:10 2022] import_file_process.pl: Global symbol "%stats_ref" requires explicit package name (did you forget to declare "my %stats_ref"?) at lib/ProductOpener/Import.pm line 524.
```

Reply:
Perl passes arguments as a list of scalars. The prototype is also coercing the arguments to scalars with the `$` sigil anyway.

Try this:

``` perl
use Data::Dumper;

sub get_org_id_for_product($$$$$$) {

    my $imported_product_ref = shift;
    my $org_id               = shift;
    my $org_ref              = shift;
    my %stats                = shift;
    my $args_ref             = shift;
    my $glns_ref             = shift;

    print Data::Dumper->Dump(
        [   $imported_product_ref,
            $org_id, $org_ref, \%stats, $args_ref, $glns_ref
        ],
        [qw(imported_product_ref org_id org_ref *stats args_ref glns_ref)]
    );
}

my %stats = (
    'products_in_file'                 => {},
    'products_already_existing'        => {},
    'products_created'                 => {},
    'products_data_updated'            => {},
    'products_data_not_updated'        => {},
    'products_info_added'              => {},
    'products_info_changed'            => {},
    'products_info_updated'            => {},
    'products_info_not_updated'        => {},
    'products_nutrition_added'         => {},
    'products_nutrition_changed'       => {},
    'products_nutrition_updated'       => {},
    'products_nutrition_not_updated'   => {},
    'products_images_added'            => {},
    'products_with_images'             => {},
    'products_with_data'               => {},
    'products_with_info'               => {},
    'products_with_ingredients'        => {},
    'products_with_nutrition'          => {},
    'products_with_nutrition_prepared' => {},
    'products_without_images'          => {},
    'products_without_data'            => {},
    'products_without_info'            => {},
    'products_without_info'            => {},
    'products_without_nutrition'       => {},
    'products_updated'                 => {},

    # Keep track of the number of products that are not imported
    # because the source does not have authorization for the org
    'orgs_without_source_authorization' => {},
    'orgs_created'                      => {},
    'orgs_existing'                     => {},
    'orgs_in_file'                      => {},
);

get_org_id_for_product( "product_ref", "org_id", "org_ref", %stats,
    "args_ref", "glns_ref" );
```

Output:
``` shell
Odd number of elements in hash assignment at hashpass.pl line 11.
$imported_product_ref = 'product_ref';
$org_id = 'org_id';
$org_ref = 'org_ref';
%stats = (
           '29' => undef
         );
$args_ref = 'args_ref';
$glns_ref = 'glns_ref';
```

As you can see, it's not working fine because `%stats` gets evaluated in scalar context, resulting in the number of items in the hash.
You need to take reference of the hash with `\` on the caller side and change the subroutine to work with the reference instead

``` perl
...
my $stats_ref = shift # Note the sigil $, references are scalars.
...
```

The rest of the subroutine also needs to be modified to work with the reference instead of the hash.